zookeeper-server/src/main/java/org/apache/zookeeper/ClientCnxn.java

```diff
@@ -971,10 +980,20 @@ void readResponse(ByteBuffer incomingBuffer) throws IOException {
          *
          * @return
          */
-        ZooKeeper.States getZkState() {
+        synchronized ZooKeeper.States getZkState() {
             return state;
```

Comment:
`state` is already modified by `volatile`, so `synchronized` is no need ?

Reply:
Thank you for your comments.
In my memory, checkstyle may report exception if remove `synchronized` from getkState. But I can not reproduce the checkstyle error in current version. Maybe I missed something. 
Another reason of adding `synchronized` is to prevent future risk.
In current implementation, we have implicit convention for setZkSate() - state modification must be placed at the end of setZkState() otherwise incomplete internal state may be exposed to outside. Adding `synchronized` to getZkState() can avoid this risk because getZkState() only be allowed to interleave execution with the whole setZkState().

Anyway I accept your suggestion. Maybe it has potential performance benefit if I remove `synchronized` from getZkState(). I will add more comments to clarify the implicit convention of state modification.

Reply:
I actually prefer the previous synchronized getZkState() approach - it's simple and consistent and easy to understand than the current approach. It's good to have performance in mind, but unless there is a real bottleneck, I would always choose simple code over premature optimization.

Reply:
Thank you for your comments.
You convince me. Accept it. Add `synchronized` back to getZkState()